### DISCUSSION TITLE
M5StackでSDカードに格納されているMP3を流せない

### DISCUSSION THREAD

**Discussion by daizusoysoy** [1ft-seabass/ma-hackathon-qa-and-knowledge-202210#19] [Q&A] [answered]
[ESP8266Audio](https://github.com/earlephilhower/ESP8266Audio)ライブラリをDLし、
SDカード直下に置かれているMP3ファイルを流そうとしています。

ネット記事に挙がっておりました下記コードを使用していますが、
ビーーというノイズ音がずっと流れるだけで、曲が流れません。
1秒のMP3ファイル（音声）と４分台のMP3ファイル（曲）両方で試しましたが、両方ともノイズとなっております。
ノイズ音の長さはMP3ファイルごとで変わっているため、読み込んではいると思うのですが、ノイズ音が解消しません。
曲が流れない理由や対処方法をご教授頂けますと幸いでございます。

```
#include <M5Stack.h>
#include <WiFi.h>
#include "AudioFileSourceSD.h"
#include "AudioFileSourceID3.h"
#include "AudioGeneratorMP3.h"
#include "AudioOutputI2S.h"


AudioGeneratorMP3 *mp3;
AudioFileSourceSD *file;
AudioOutputI2S *out;
AudioFileSourceID3 *id3;

void setup()
{
  M5.begin();
  WiFi.mode(WIFI_OFF); 
  delay(500);

  M5.Lcd.setTextFont(2);
  M5.Lcd.printf("Sample MP3 playback begins...\n");
  Serial.printf("Sample MP3 playback begins...\n");
}

void playMP3(char *filename){
  file = new AudioFileSourceSD(filename);
  id3 = new AudioFileSourceID3(file);
  out = new AudioOutputI2S(0, 1); // Output to builtInDAC
  out->SetOutputModeMono(true);
  out->SetGain(1.0);
  mp3 = new AudioGeneratorMP3();
  mp3->begin(id3, out);
  while(mp3->isRunning()) {
    if (!mp3->loop()) mp3->stop();
  }
}

void loop()
{
  M5.update();
  if(M5.BtnA.isPressed()){
    playMP3("/voice1.mp3");
  }
  if(M5.BtnB.isPressed()){
    playMP3("/voice1.mp3");
  }
  if(M5.BtnC.isPressed()){
    playMP3("/sample.mp3");
  }
  Serial.printf("MP3 done\n");
  delay(10);

}
```

**Comment by 1ft-seabass** [comment, marked as answer]
@daizusoysoy こちら試してみますー。

**Reply by 1ft-seabass** [reply]
@daizusoysoy なるほど！これは爆音ノイズですね！再現はできたので、調べていきます。

**Reply by daizusoysoy** [reply]
ありがとうございます。何卒よろしくお願いいたします。

**Reply by 1ft-seabass** [reply]
M5Stackのスピーカーの音質が悪い原因と対策 - Qiita
https://qiita.com/Tw_Mhage/items/34be8e8b42580c449b56

M5StackでSDカードからmp3を再生してみる - Qiita
https://qiita.com/ktansai/items/f0096495e2ca7fa38eb7

このあたりがヒントになりそうです。わたしもちょっと試してみます。 @daizusoysoy

**Reply by 1ft-seabass** [reply]
1秒程度無音を入れても改善しなかったです。ちょっとゲインが怪しいですね。

**Reply by 1ft-seabass** [reply]
out->SetGain(0.001); でメッチャ小さくしたんですが、ソフト側の制御だと変わらないようです。となると、音源自体のゲインを抑えるのがいいかもです。

**Reply by 1ft-seabass** [reply]
@daizusoysoy

**Reply by 1ft-seabass** [reply]
うーん、ここで言及されている Piano ファイルでもノイズのりますね。何かコードに違いがないか見てみます。
https://qiita.com/hextomino/items/906da42151d5ad545a60

**Reply by daizusoysoy** [reply]
お疲れ様です。MP3自体の音量を下げたものと、下げてないものとで比較しましたが特にノイズもノイズ音量の違いも見受けられませんでした。

**Reply by 1ft-seabass** [reply]
しいていえば、out->SetGain(1.0);なんですが、これ抜いてもおかしいですね。

**Reply by 1ft-seabass** [reply]
@daizusoysoy あ、ちょっとわかってきたかも

**Reply by 1ft-seabass** [reply]
@daizusoysoy いま、ちょっとコードの先入観を捨てて、冷静にコード見てます。

**Reply by 1ft-seabass** [reply]
@daizusoysoy うーんダメでした。
https://qiita.com/hextomino/items/906da42151d5ad545a60

このコードのまま動かしてもノイズが乗りますねー。

実は一番最初の @daizusoysoy さんのコードはボタン押すたびに while 文が多重に走る可能性を感じてて、これかなって思ったんですが違いました。

**Reply by 1ft-seabass** [reply]
@daizusoysoy 
音声再生 - MP3再生 - m5stackで組込み！！
https://m5stack-build.hatenablog.com/entry/2019/12/21/225908

これが良さそうなの追ってみます。

**Reply by daizusoysoy** [reply]
連携いただいているURLはすべていったん試していますが、どれもノイズになってしまいますね…

**Reply by 1ft-seabass** [reply]
@daizusoysoy ですよねー。M5Stack V2.6 特有でおかしいのか、Arduino IDE の特定バージョンがおかしいのか・・・？もう少し調べてみます。

とりあえず、音系ハマりそうなので、他の制作も進めつつでお願いしますー。

**Reply by 1ft-seabass** [reply]
バージョンで何か影響あるかもなので調べてみます。

https://twitter.com/mongonta555/status/1541600606091374592

**Reply by daizusoysoy** [reply]
なかなか解決せず申し訳ないです。
明日朝段階で解決していなければこの機能は削除を考えております。
何卒宜しくお願い致します。

**Reply by 1ft-seabass** [reply]
@daizusoysoy こちらもなかなか解決できずすいません。引き続き調べていますが、進捗があんまりかんばしくないので、おそれいりましすが削除も視野に入れつつ動いていて抱ければと思います。

**Reply by 1ft-seabass** [reply]
M5STACK（その8）
http://nekomemo2.site/?eid=1569

ビットレートまわりをちょっと怪しんでみます。

**Reply by 1ft-seabass** [reply]
ピアノのを 48kbps 固定ビットレートでモノラル強制でやったんですがノイズで同様。うまく動きませんでした。

**Reply by 1ft-seabass** [reply]
M5Stack/PlayMP3FromSDToDAC.ino at master · m5stack/M5Stack
https://github.com/m5stack/M5Stack/blob/master/examples/Advanced/Audio/PlayMP3FromSDToDAC/PlayMP3FromSDToDAC.ino

M5Stack 純正のコードでノイズが走るので、ちょっと困りますねー。追います。

**Reply by 1ft-seabass** [reply]
@daizusoysoy いろいろ検証したのですが

- 別の M5Stack V2.6 でも同様にノイズが出る
- M5Stack Gray でも同様にノイズが出る
- 2022年6月ごろに成功したナレッジがある

といったところで、ESP8266Audioなどのライブラリのバージョンがあやしいのですが、ここを検証するとなると、いくつかのライブラリのバージョンを用意して検証することとなり、今回のハッカソンでしっかりした知見に落ち着かなさそうです。

なので、今回は選択しないほうが良さそうです。明確な解決策が示せずすいませんが、よろしくお願いします。

**Reply by daizusoysoy** [reply]
ご検証いただきましてありがとうございました。
昨日自分のほうでも、ESP8266AudioのバージョンをM5Stackのバージョンに近しい数値のものに変更して試したりもしてみたのですが、ダメそうでしたので今回は見送ろうと思います。

お忙しいところありがとうございました。
